client/src/main/java/io/apicurio/registry/rest/client/impl/RegistryClientImpl.java

```diff
@@ -0,0 +1,303 @@
+
+		try {
+			final Map<String, String> params = Map.of(Parameters.LIMIT, String.valueOf(limit), Parameters.OFFSET, String.valueOf(offset), Parameters.SORT_ORDER, order.value(), Parameters.ORDER_BY, orderby.value());
+
+			final HttpRequest req = HttpRequest.newBuilder()
+					.uri(buildURI(endpoint + Routes.GROUP_BASE_PATH, params, groupId))
+					.GET()
+					.build();
+
+			final HttpResponse<InputStream> res = client.send(req, BodyHandlers.ofInputStream());
+
+			if (200 == res.statusCode()) {
+				return mapper.readValue(res.body(), ArtifactSearchResults.class);
+			}
+		} catch (URISyntaxException | IOException | InterruptedException e) {
+			throw parseError(e);
+		}
+		return null;
+	}
+
+	private RestClientException parseError(Exception ex) {
+
+		//TODO build error
+		return new RestClientException(new Error());
+	}
+
+	@Override
+	public ArtifactMetaData createArtifact(String groupId, ArtifactType xRegistryArtifactType, String xRegistryArtifactId, String xRegistryVersion, IfExists ifExists, Boolean canonical, InputStream data) {
+
+		try {
+
+			final HttpRequest req = HttpRequest.newBuilder()
+					.header(Headers.ARTIFACT_ID, xRegistryArtifactId)
+					.header(Headers.ARTIFACT_TYPE, xRegistryArtifactType.value())
+					.header(Headers.VERSION, xRegistryVersion)
+					.uri(buildURI(endpoint + Routes.GROUP_BASE_PATH, Collections.emptyMap(), groupId))
+					.POST(HttpRequest.BodyPublishers.ofByteArray(data.readAllBytes()))
+					.build();
+
+			return client.send(req, new JsonBodyHandler<>(ArtifactMetaData.class))
+					.body()
+					.get();
+
+		} catch (URISyntaxException | IOException | InterruptedException e) {
+			throw parseError(e);
+		}
+	}
+
+	@Override
+	public void deleteArtifactsInGroup(String groupId) {
+
+	}
+
+	@Override
+	public InputStream getContentById(int contentId) {
+		return null;
+	}
+
+	@Override
+	public InputStream getContentByGlobalId(int globalId) {
+		return null;
+	}
+
+	@Override
+	public InputStream getContentByHash(int contentHash, Boolean canonical) {
+		return null;
+	}
+
+	@Override
+	public ArtifactSearchResults searchArtifacts(String name, Integer offset, Integer limit, SortOrder order, SortBy orderby, List<String> labels, List<String> properties, String description, String artifactgroup) {
+		return null;
+	}
+
+	@Override
+	public ArtifactSearchResults searchArtifactsByContent(Integer offset, Integer limit, SortOrder order, SortBy orderby, InputStream data) {
+		return null;
+	}
+
+	private static URI buildURI(String basePath, Map<String, String> queryParams, Object... pathParams) throws URISyntaxException {
+
+		final URIBuilder uriBuilder = new URIBuilder(String.format(basePath, pathParams));
+
+		uriBuilder.setParameters(queryParams.keySet()
+				.stream()
+				.map(key -> new BasicNameValuePair(key, queryParams.get(key)))
+				.collect(Collectors.toList())
+		);
```

Comment:
Cool.  Although it would be better to use `.entrySet()` here instead of `.keySet()` - that would save you another lookup.

Reply:
Oh my...